### PR TITLE
Minor perf improvement to validator

### PR DESCRIPTION
**Pull request by codyebberson**
@mattwiller - for your consideration.  see comments inline

### CODE REVIEW

File: packages/core/src/typeschema/validation.ts [hunk 1]
```diff
@@ -316,27 +314,6 @@ function isChoiceOfType(
   return false;
 }
 
-function getNestedProperty(value: TypedValue, key: string): (TypedValue | TypedValue[] | undefined)[] {
-  const [firstProp, ...nestedProps] = key.split('.');
```
Comment (codyebberson) [inline packages/core/src/typeschema/validation.ts:320]: I ran a sanity test on this  `if (key.includes('.')) throw new Error('period');` and it didn't throw

Up above, it looks like you're already doing the key splitting and path management.  This function never enters into the `for` loop.

File: packages/core/src/typeschema/validation.ts [hunk 2]
```diff
@@ -113,46 +113,44 @@ class ResourceValidator {
   }
 
   private checkProperty(parent: TypedValue, key: string, schema: InternalTypeSchema, path: string): void {
-    const propertyValues = getNestedProperty(parent, key);
+    const value = getTypedPropertyValue(parent, key);
```
Comment (codyebberson) [inline packages/core/src/typeschema/validation.ts:116]: By replacing `getNestedProperty` with a direct call to `getTypedPropertyValue`, you can avoid the outer array, and avoid the for loop on old line 121.

**Review by reshmakh — APPROVED** [review]

**Comment by mattwiller** [comment]
The case that `getNestedPropertyValue()` was intended to handle was when profiles "walk into" types to place constraints on specific fields; for example, the [US Core Patient profile](http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient) makes `Patient.telecom.system` required.  Let me try adding a test case for that to ensure it's still working correctly

**Review by mattwiller — CHANGES_REQUESTED** [review]
I added a test case on this branch with a Patient that almost conforms to the US Core Patient profile:
```js
{
  resourceType: 'Patient',
  identifier: [{
    system: 'http://example.com',
    value: 'foo',
  }],
  telecom: [{
    // Missing system property
    value: '555-555-5555',
  }],
  gender: 'unknown',
  name: [{
      given: ['Test'],
      family: 'Patient',
  }],
}
```
Unfortunately it fails:
```
Expected message: "Missing required property (Patient.telecom.system)"
Received message: "Missing required property (Patient.identifier.system); Missing required property (Patient.identifier.value); Missing required property (Patient.telecom.system); Missing required property (Patient.telecom.value)"
```
I verified that this same test passes on `main` and will create a PR to add it

File: packages/core/src/typeschema/validation.test.ts [hunk 3]
```diff
@@ -790,16 +790,6 @@ describe('Legacy tests for parity checking', () => {
     }
   });
 
-  test('StructureDefinition', () => {
```
Comment (codyebberson) [inline packages/core/src/typeschema/validation.test.ts:793]: I believe this test is identical to the test on line 318

Comment (mattwiller) [reply packages/core/src/typeschema/validation.test.ts:793]: It is, great catch!

**Comment by codyebberson** [comment]
> Unfortunately it fails:
> 
> ```
> Expected message: "Missing required property (Patient.telecom.system)"
> Received message: "Missing required property (Patient.identifier.system); Missing required property (Patient.identifier.value
> ```

Interesting.  Ok, I'll take a look at this.  Are "expected" and "received" reversed?  It seems like "received" is a superset of expected?  or are those other errors wrong?

**Comment by mattwiller** [comment]
>  Are "expected" and "received" reversed? It seems like "received" is a superset of expected? or are those other errors wrong?

Nope, that's the correct labeling: the only error in the test resource should be the absence of `Patient.telecom.system`

**Comment by codyebberson** [comment]
@mattwiller  Thanks for new test case! Makes more sense now.  Closing this out.

**Comment by mattwiller** [comment]
I think there is still a possibility that we could get rid of the nested property accesses, but we might need to tweak the `InternalTypeSchema` representation to make that possible

**Pull request closed without merging**